lib/graphql/internal_representation/rewrite.rb

```diff
@@ -138,6 +144,12 @@ def validate(context)
           # Pop this field's node
           # and record any directives that were visited
           # during this field & before it (eg, inline fragments)
+          
+          return if ast_node.directives.any? do |directive|
+            directive.name == 'skip' && directive.arguments.any? {|arg| arg.value == true } ||
+            directive.name == 'include' && directive.arguments.any? {|arg| arg.value == false }
+          end
```

Comment:
Yes - without shortcircuiting the leave hook, the `@nodes` stack gets out of sync with the current location in the AST, which breaks fragment resolution later on.